torchdata/datapipes/iter/transform/callable.py

```diff
@@ -168,6 +171,146 @@ def __len__(self) -> int:
+        input_col: Index or indices of data which ``fn`` is applied, such as:
+
+            - ``None`` as default to apply ``fn`` to the data directly.
+            - Integer(s) is/are used for list/tuple.
+            - Key(s) is/are used for dict.
+        buffer_size: The buffer size for sampling the iterables (default to ``10000``)
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> source_dp = IterableWrapper([[1, 2, 3, 4], 'abcd', 'ABCD'])
+        >>> shuffled_flatmapped_dp = source_dp.shuffled_flatmap(buffer_size=2)
+        >>> list(shuffled_flatmapped_dp)
+        ['a', 'b', 'c', 1, 'd', 'A', 'B', 'C', 2, 'D', 3, 4]
+        >>>
+        >>> # To shuffle all the elements, you can combine `shuffled_flatmap` with `in_batch_shuffle` like this:
+        >>> fully_shuffled_flatmapped_dp = source_dp.in_batch_shuffle()
+        >>> fully_shuffled_flatmapped_dp = fully_shuffled_flatmapped_dp.shuffled_flatmap()
+        >>> list(fully_shuffled_flatmapped_dp)
+        ['b', 3, 'c', 'd', 'C', 'A', 'a', 2, 'B', 'D', 4, 1]
+    """
+    datapipe: IterDataPipe[T_co]
+    fn: Optional[Callable]
+    buffer_size: int
+    _enabled: bool
+    _seed: Optional[int]
+    _rng: random.Random
+
+    def __init__(
+        self, datapipe: IterDataPipe, fn: Optional[Callable] = None, input_col=None, buffer_size: int = 100
+    ) -> None:
+        super().__init__()
+        self.datapipe = datapipe
+
+        if fn is None:
+            fn = _no_op_fn
+        _check_unpickable_fn(fn)
+        self.fn = fn  # type: ignore[assignment]
+        self.input_col = input_col
+        validate_input_col(fn, input_col)
+
+        assert buffer_size > 0, "buffer_size should be larger than 0"
+        self.buffer_size = buffer_size
+        self._enabled = True
+        self._seed = None
+        self._rng = random.Random()
+
+    def set_shuffle(self, shuffle=True):
+        self._enabled = shuffle
+        return self
+
+    def set_seed(self, seed: int):
+        self._seed = seed
+        return self
+
+    def reset(self) -> None:
+        if self._enabled:
+            if self._seed is None:
+                self._seed = int(torch_empty((), dtype=torch_int64).random_().item())
+            self._rng.seed(self._seed)
+            self._seed = None
+
+    def _apply_fn(self, data):
+        if self.input_col is None:
+            return self.fn(data)  # type: ignore[misc]
+        elif isinstance(self.input_col, (list, tuple)):
+            args = tuple(data[col] for col in self.input_col)
+            return self.fn(*args)  # type: ignore[misc]
+        else:
+            return self.fn(data[self.input_col])  # type: ignore[misc]
+
+    def _yield_one(self, buffer: List[Iterator[T_co]]) -> Iterator[T_co]:
+        idx = self._rng.randint(0, len(buffer) - 1) if self._enabled else 0
+        try:
+            yield next(buffer[idx])
+        except StopIteration:
+            buffer.pop(idx)
```

Comment:
Yes I see, but it is not easy to ensure that only one random `idx` is generated per element yielded: when you finished iterating over the elements of the input datapipe (so you are only emptying the buffer, not adding new elements to it), the `idx` you sampled might not exist in the buffer after the `pop`(s).   

I changed the code so that we don't sample unnecessary `idx` while we are iterating over the input data pipe. Would this solution be ok: https://github.com/PierreGtch/TorchData/commit/97f2aeb2b5b709c441ed3337286acdd7cde5d389 

Otherwise, we can keep an additional buffer containing the next element of every iterator in the buffer to ensure that only one `idx` per item is sampled.